scripts/dts-localize.ts

```diff
@@ -0,0 +1,261 @@
+  if (!path.isAbsolute(fileName)) {
+    fileName = path.join(cwd !== undefined ? cwd : process.cwd(), fileName);
+  }
+
+  return fileName;
+}
+
+function getCompilerOptions(inputFileNames: string[], preferredConfigPath?: string) {
+  const configFileName =
+    preferredConfigPath !== undefined ? preferredConfigPath : findConfig(inputFileNames);
+  const configParseResult = ts.readConfigFile(configFileName, ts.sys.readFile);
+  const compilerOptionsParseResult = ts.parseJsonConfigFileContent(
+    configParseResult.config,
+    parseConfigHost,
+    path.resolve(path.dirname(configFileName)),
+    undefined,
+    getAbsolutePath(configFileName)
+  );
+
+  return compilerOptionsParseResult.options;
+}
+
+function findConfig(inputFiles: string[]) {
+  if (inputFiles.length !== 1) {
+    throw new Error(
+      'Cannot find tsconfig for multiple files. Please specify preferred tsconfig file'
+    );
+  }
+
+  // input file could be a relative path to the current path
+  // and desired config could be outside of current cwd folder
+  // so we have to provide absolute path to find config until the root
+  const searchPath = getAbsolutePath(inputFiles[0]);
+  const configFileName = ts.findConfigFile(searchPath, ts.sys.fileExists);
+  if (!configFileName) {
+    throw new Error(`Cannot find config file for file ${searchPath}`);
+  }
+
+  return configFileName;
+}
+
+interface Options {
+  externals: string[];
+  cwd?: string;
+}
+
+export const run = async (entrySourceFiles: string[], outputPath: string, options: Options) => {
+  const compilerOptions = getCompilerOptions(entrySourceFiles);
+  const host = ts.createCompilerHost(compilerOptions);
+  const cwd = options.cwd || process.cwd();
+  const pkg = sync({ cwd }).packageJson;
+  const externals = Object.keys({ ...pkg.dependencies, ...pkg.peerDependencies });
+
+  // this to make paths for local packages as they are in node_modules because of yarn
+  // but it depends on the way you handle "flatting of files"
+  // so basically you can remove this host completely if you handle it in different way
+  host.realpath = (p: string) => p;
+
+  const program = ts.createProgram(entrySourceFiles, compilerOptions, host);
+  const printer = ts.createPrinter({ newLine: ts.NewLineKind.LineFeed, removeComments: false });
+
+  const typeChecker = program.getTypeChecker();
+  const sourceFiles = program.getSourceFiles();
+
+  const filesRemapping = new Map<string, string>();
+  const replaceRemapping = new Map<string, string[]>();
+
+  for (const sourceFile of sourceFiles) {
+    // skip lib files from the compiler
+    if (program.isSourceFileDefaultLibrary(sourceFile)) {
+      continue;
+    }
+
+    // here you can implement any logic regarding how your new files should be placed
+    // as for this POC I used rough solution to use the same structure but I believe it should be improved
+    filesRemapping.set(sourceFile.fileName, getNewPath(cwd, sourceFile.fileName));
```

Comment:
Is it necessary to do the mapping for all source files up front? As you walked me through it it seemed like instead we could do it on demand. I'm not sure how many files we are talking about or how expensive it is, perhaps that's a premature optimisation.